[sw,cryptolib] Cryptolib p384 add blinding refresh to point multiplication

h-filali opened the pull request:
This commit adds refreshing to the blinding of the secret key scalar d before
every execution of the point multiplication.

Furthermore this PR addresses an issue in the existing point multiplication for which
the src and dst of a `bn.sel` instruction are the same register.

For further info please have a look at the individual commit messages.

Related to https://github.com/lowRISC/opentitan/issues/27648.

andrea-caforio commented on sw/otbn/crypto/p384_internal_mult.s:
Why it is 65 bits needs to be explained.

andrea-caforio commented on sw/otbn/crypto/p384_internal_mult.s:
Why 194 bits?

andrea-caforio reviewed (commented):
If I understood correctly, there are now two multiplication routines `scalar_mult_int_p384` and `scalar_mult_int_p384_reblind` with latter having
an additional indirection via `p384_masked_scalar_hide` that reblinds the
scalar `d`. Reblinding is not needed for some applications but I'm not sure if
having two large almost identical routines is the way to go. Can it be refactored
differently as to avoid this heavy code duplication?

andrea-caforio commented:
By the way we should move away from using `.exp` files in unit tests and use
@sasdf new testcase feature which is a superior approach that streamlines
tests more easily while avoiding having to store test data alongside application
data in DMEM.

h-filali commented:
Thanks @andrea-caforio I addressed your comments.

Regarding the `scalar_mult_int_p384` and `scalar_mult_int_p384_reblind` routines, those have some differences like the loop iterations which make it a bit hard to factor into a third function. And unfortunately both routines are needed if we don't want to take a penalty on execution time for the random scalar multiplication for sign.

 I also renamed `scalar_mult_int_p256_reblind` in https://github.com/lowRISC/opentitan/pull/27900 to conform more with the naming in this PR.
 
 Lastly, I changed the tests to use the hjson framework.
 
 Would you mind having another look?

johannheyszl commented on sw/otbn/crypto/p384_internal_mult.s:
```diff
@@ -332,16 +332,16 @@ scalar_mult_int_p384:
-    bn.sel    w2, w2, w10, C
-    bn.sel    w3, w3, w11, C
+      [w3, w2] = z <= [w1, w0] mod p */
+    bn.sub    w10, w0, w12
```
should this be w2 and w3 @h-filali ?

johannheyszl commented on sw/otbn/crypto/p384_internal_mult.s:
typo offeset

johannheyszl commented on sw/otbn/crypto/p384_internal_mult.s:
does our testing cover this, i.e. would we detect bugs in the multiplication?

h-filali commented on sw/otbn/crypto/p384_internal_mult.s:
Replaced it with RND now to be safe!

h-filali commented:
@johannheyszl I should have addressed all your comments. Thanks for reviewing. I also created a separate function of the main body of the internal multiplication. This should save some code size. 

Regarding the execution times using a single main body function we go from:
- Keygen: 1857674
- Keygen side loaded: 1857674
- ECDH: 1869275
- ECDH side loaded: 1869411
- Sign: 1505666
- Sign side loaded: 1505802

to:
- Keygen: 1857677
- Keygen side loaded: 1857585
- ECDH: 1869278
- ECDH side loaded: 1869414
- Sign: 1508810
- Sign side loaded: 1508946

These numbers are on top of the instruction counts I commented above.

johannheyszl commented on sw/otbn/crypto/p384_internal_mult.s:
```diff
@@ -232,7 +343,7 @@ scalar_mult_int_p384:
 
   /* double-and-add loop with decreasing index */
-  loopi     448, 85
+  loop      x16, 92
```
can we please add a check of x16 whether it is either 448 or 578 and through an error otherwise?

h-filali commented:
<img width="2687" height="1474" alt="image" src="https://github.com/user-attachments/assets/8b7cd1ae-caba-4461-b85c-8af456b40342" />
Here's the new map file analysis. Maybe lets give people the chance to have a look and then merge if everything is fine.

nasahlpa commented:
> <img alt="image" width="2000" height="1474" src="https://private-user-images.githubusercontent.com/93996688/493345618-8b7cd1ae-caba-4461-b85c-8af456b40342.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3MTk4MjIsIm5iZiI6MTc1ODcxOTUyMiwicGF0aCI6Ii85Mzk5NjY4OC80OTMzNDU2MTgtOGI3Y2QxYWUtY2FiYS00NDYxLWI4NWMtOGFmNDU2YjQwMzQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI0VDEzMTIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzMjcxZGM3MDFjNGZlY2UyNThmMGYwY2MwZGNjMzcxNmRlMTE5YzUzMWEwYjVlZDk3OGE2NTdlZTJlOTU4NWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gXnN9wpb8EAHes95bsN1dnTHfCuzHhXtvLndtthlSEw"> Here's the new map file analysis. Maybe lets give people the chance to have a look and then merge if everything is fine.

Can you please post the code size before and after your change?

h-filali commented:
> > <img alt="image" width="2000" height="1474" src="https://private-user-images.githubusercontent.com/93996688/493345618-8b7cd1ae-caba-4461-b85c-8af456b40342.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3MTk4MjIsIm5iZiI6MTc1ODcxOTUyMiwicGF0aCI6Ii85Mzk5NjY4OC80OTMzNDU2MTgtOGI3Y2QxYWUtY2FiYS00NDYxLWI4NWMtOGFmNDU2YjQwMzQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI0VDEzMTIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzMjcxZGM3MDFjNGZlY2UyNThmMGYwY2MwZGNjMzcxNmRlMTE5YzUzMWEwYjVlZDk3OGE2NTdlZTJlOTU4NWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gXnN9wpb8EAHes95bsN1dnTHfCuzHhXtvLndtthlSEw"> Here's the new map file analysis. Maybe lets give people the chance to have a look and then merge if everything is fine.
> 
> Can you please post the code size before and after your change?

Sure thats before
<img width="2687" height="1474" alt="image" src="https://github.com/user-attachments/assets/71e8c93e-1b1c-4be1-8a69-5316f6f4d179" />

nasahlpa reviewed (approved):
Thanks Hakim, I had a closer look into the new point multiplication.
